mmaction/models/heads/ssn_head.py

```diff
@@ -0,0 +1,417 @@
+import torch
+import torch.nn as nn
+from mmcv.cnn import normal_init
+
+from ..registry import HEADS
+
+
+def parse_stage_config(stage_cfg):
+    """Parse config of STPP for three stages.
+
+    Args:
+        stage_cfg (int | tuple[int]):
+            Config of structured temporal pyramid pooling.
+
+    Returns:
+        tuple[tuple[int], int]:
+            Config of structured temporal pyramid pooling and
+            total number of parts(number of multipliers).
+    """
+    if isinstance(stage_cfg, int):
+        return (stage_cfg, ), stage_cfg
+    elif isinstance(stage_cfg, tuple):
+        return stage_cfg, sum(stage_cfg)
+    else:
+        raise ValueError(f'Incorrect STPP config {stage_cfg}')
+
+
+class STPPTrain(nn.Module):
+    """Structured temporal pyramid pooling for SSN at training.
+
+    Args:
+        stpp_stage (tuple): Config of structured temporal pyramid pooling.
+            Default: (1, (1, 2), 1).
+        num_segments_list (tuple): Number of segments to be sampled
+            in three stages. Default: (2, 5, 2).
+    """
+
+    def __init__(self, stpp_stage=(1, (1, 2), 1), num_segments_list=(2, 5, 2)):
+        super().__init__()
+
+        starting_part, starting_multiplier = parse_stage_config(stpp_stage[0])
+        course_part, course_multiplier = parse_stage_config(stpp_stage[1])
+        ending_part, ending_multiplier = parse_stage_config(stpp_stage[2])
+
+        self.num_multipliers = (
+            starting_multiplier + course_multiplier + ending_multiplier)
+        self.stpp_stages = (starting_part, course_part, ending_part)
+        self.multiplier_list = (starting_multiplier, course_multiplier,
+                                ending_multiplier)
+
+        self.num_segments_list = num_segments_list
+
+    def forward(self, x, scale_factors):
+        """Defines the computation performed at every call.
+
+        Args:
+            x (torch.Tensor): The input data.
+            scale_factors (list): Ratios of the effective sampling lengths
+                to augmented lengths.
+
+        Returns:
+            tuple[torch.Tensor, torch.Tensor]:
+                Features for predicting activity scores and
+                completeness scores.
+        """
+        x0 = self.num_segments_list[0]
+        x1 = x0 + self.num_segments_list[1]
+        num_segments = x1 + self.num_segments_list[2]
+
+        feat_dim = x.size(1)
+        x = x.view(-1, num_segments, feat_dim)
+        num_samples = x.size(0)
+
+        scale_factors = scale_factors.view(-1, 2)
+
+        def extract_stage_feature(stage_feat, stage_parts, num_multipliers,
```

Comment:
It can be set as a member function to avoid the overhead by using function in function.
```
def _extract_stage_feature(self, xxxx)
```